Allow relationship ids to be reused

MishaDemianenko opened the pull request:
Introduce feature toggle `unsupported.dbms.relationships.ids.reuse` to enable relationship id reuse.
Id reuse will be handled by standard mechanism where relationship id should be marked as free only when there is no active transaction overlap.

MishaDemianenko committed:
Allow relationship ids to be reused

Introduce feature toggle `unsupported.dbms.relationships.ids.reuse` to enable relationship id reuse.
Id reuse will be handled by standard mechanism where relationship id should be marked as free only when there is no active transaction overlap.

MishaDemianenko committed:
Add simple scenarios tests for relationship cursors.

Add missing tests for relationship cursors simple scenarios.

MishaDemianenko committed:
Do maintenance on buffering id generator when closing it.

Perform additional maintenance when closing buffering id generator to release ids that where in a buffer since last cleanup and satisfy release criteria.

MishaDemianenko committed:
Introduce separate component for id maintenance and put it to dependency resolver.
Add stress test for concurrent relationship creation/deletion/read.

MishaDemianenko committed:
High level id reuse integration test

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/IdType.java:
```diff
@@ -19,14 +19,16 @@
 public enum IdType
 {
     NODE( 35, false ),
-    RELATIONSHIP( 35, false ),
```
Usually FeatureToggles doesn't look like usual settings. I'm leaning more towards having this as a normal `Setting` though.

MishaDemianenko replied:
problem with old plain `Setting` is that we need to apply it to some `Config` and pass it around, that even more trickier because it's enum, do you really want it to be standard `Setting`?

tinwelint replied:
Yes, I think this is a "normal" feature, not a special hidden thing. Instead of doing this inside the IdType enum the config thing could be done in the place using that value, right?

tinwelint commented on community/kernel/src/test/java/org/neo4j/graphdb/IdReuseTest.java:
```diff
@@ -65,15 +73,15 @@ public void shouldReuseRelationshipIdsFromRolledBackTransaction() throws Excepti
         // Given
         GraphDatabaseService db = dbRule.getGraphDatabaseService();
         Node node1, node2;
-        try (Transaction tx = db.beginTx())
```
This is all good and fixing checkstyle anomalies, but to do it in maintenance branches causes more harm than good usually.

MishaDemianenko replied:
sorry for that, updated

MishaDemianenko committed:
Allow id type to be configurable, edition based configuration of id type.

Allow id type reusability to be configurable during database edition configuration.
Introduce enterprise edition specific configurator and settings to be able to disticnt specific settings
and perform custom configuration.
Allow relationship ids to be reusable in ha and enterprise editions.

tinwelint commented:
Marking as not ready for merge since it awaits a solution around reading on HA slaves before able to go in. Also this needs a rebase @MishaDemianenko

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/BufferingIdGenerator.java:
```diff
@@ -70,4 +70,14 @@ void maintenance()
+    {
+        if ( buffer != null )
+        {
+            buffer.maintenance();
```
Rather we should call `buffer.close()` since it has one and does additional closy stuff there.

MishaDemianenko replied:
true :)

tinwelint commented on community/consistency-check-legacy/src/main/java/org/neo4j/legacy/consistency/checking/AbstractStoreProcessor.java:
```diff
@@ -144,25 +143,28 @@ public final void processString( RecordStore<DynamicRecord> store, DynamicRecord
         RecordType type;
         DynamicStore dereference;
-        switch ( idType )
+        if ( IdType.STRING_BLOCK.equals( idType ) )
```
Why make this change (a) in a maint branch and (b) at all?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/BufferingIdGeneratorFactory.java:
```diff
@@ -32,8 +35,7 @@
 {
-    private final BufferingIdGenerator[/*IdType#ordinal as key*/] overriddenIdGenerators =
-            new BufferingIdGenerator[IdType.values().length];
+    private final Map<IdType, BufferingIdGenerator> overriddenIdGenerators = new HashMap<>();
```
I'd rather have the array, HashMap is way more expensive, it's an unnecessary cost. Can we introduce a typeId() on IdType instead?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/IdType.java:
```diff
@@ -19,49 +19,83 @@
  */
 @Deprecated
-public enum IdType
+public class IdType
```
To my understanding it seems that this, rather big, change w/ regards to `IdType` and all code this affects has to do with the fact that its state is now mutable/different per database, i.e. `allowAggressiveIdReuse`. I'd prefer it if we would instead separate `IdType` from its mutable state and keep `IdType` otherwise what it is. That has to be simpler, hasn't it?

tinwelint replied:
Hmm, now I don't get it actually. I though that you could have different settings for different dbs, but it still seems to be per JVM, i.e. that IdType is sort of like an enum. So what was this refactoring useful for?

MishaDemianenko replied:
Your first note was correct - all of it because of mutable state of idType.
It state will be updated during enterprise db configuration.
In case if we keeping enum to represent id type we need another class to store it state and yet another to group them together in some way - so it's even more complex right?
And i do not want to put mutable structure inside of enum because that is still mutable state

MishaDemianenko committed:
Close buffer on generator close instead of additional maintenance.

MishaDemianenko commented:
@tinwelint rebased and some comments fixed.

MishaDemianenko committed:
Refactor id type enum.
Introduce separate type for idType related configuration.
Add possibility to have different id type configurations in different editions, by supplying different idTypeConfigurationProviders.

MishaDemianenko commented:
@tinwelint id types updated

MishaDemianenko committed:
Update default settings to be compatible with current state - do not force id reuse for any type in default enterprise setup.

MishaDemianenko committed:
Add missing open method, remove error prone delegate.
Add final keywords here and there.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/CommunityIdTypeConfigurationProvider.java:
```diff
@@ -33,8 +33,8 @@
-    private Set<IdType> typesToAllowReuse = EnumSet.of( IdType.PROPERTY, IdType.STRING_BLOCK,
+    private final Map<IdType,IdTypeConfiguration> typeConfigurations = new EnumMap<>(IdType.class);
+    private final Set<IdType> typesToAllowReuse = EnumSet.of( IdType.PROPERTY, IdType.STRING_BLOCK,
             IdType.ARRAY_BLOCK, IdType.NODE_LABELS);
```
Can they also be `static`?

MishaDemianenko replied:
We will do TYPES_TO_ALLOW_REUSE static, and typeConfigurations should be instance specific

MishaDemianenko committed:
Update property description, remove maintenance method from NeoStoreDataSource, introduce constants
